In [1]:
#29/04/2026 Wednessday 
#import libraries
import pandas as pd 
import numpy as np

In [2]:
#working with numpy array 
arr = np.array([10, 20, 30, 40, 50, 60, 70])
print("1D array:", arr)
print("\nAverage number:", arr.mean())

1D array: [10 20 30 40 50 60 70]

Average number: 40.0


In [3]:
#indexing (indexing is use for accessing an element)
print(arr[3])
print(arr[-2])
print(arr[4])
print(arr[6])

40
60
50
70


In [4]:
#slicing (use to extract portion of an element using the index of the range from the object)
# for range: the last number is not inclusive
#slicing doesn't print index error
print(arr[1:6])
print(arr[6:])
print(arr[:7])
print(arr[-6:-1])
print(arr[7:])
print(arr[-1:-6])

[20 30 40 50 60]
[70]
[10 20 30 40 50 60 70]
[20 30 40 50 60]
[]
[]


In [5]:
#2D => 2 Dimensional array with indexing and slicing
#2D array contain row and column 
# axis of row is 0
# axis of column is 1
arr2 = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])


print(arr2)
print(arr2[0,1])
print(arr2[1,:])
print(arr2[1:2 , 2])
print(arr2[2, :2])
print(arr2[1:2, 0:2])

[[1 2 3]
 [4 5 6]
 [7 8 9]]
2
[4 5 6]
[6]
[7 8]
[[4 5]]


In [6]:
# 30/04/2026 Thursday 
#1D to 2D array 
arr3 = np.arange(1, 13)
print("Array numbers:", arr3)

#Converting 1D array to 2D array using "reshape"

reshape = arr3.reshape(4, 3)
print("Reshape 1D array into 2D array: \n", reshape)

#Converting 2D array to 1D array using "flatten"
flatten = reshape.flatten()
print("Flatten array:", flatten)

Array numbers: [ 1  2  3  4  5  6  7  8  9 10 11 12]
Reshape 1D array into 2D array: 
 [[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
Flatten array: [ 1  2  3  4  5  6  7  8  9 10 11 12]


In [7]:
#working with mathematics operation 
arr4 = np.array([1,2,3,4,5])
arr5 = np.array([10, 20, 30, 40, 50])

#basic math operation
print("Addition: ", arr4 + arr5)
print("Subraction: ", arr4 - arr5)
print("Multiplication:", arr4 * arr5)
print("Division:", arr4 / arr5)

#scaler operation
print("\nScaler operation \n")

print("Exponetial:", arr5 ** 2)
print("Floor division:", arr5 // 2)
print("Modulus:", arr5 % 3)

Addition:  [11 22 33 44 55]
Subraction:  [ -9 -18 -27 -36 -45]
Multiplication: [ 10  40  90 160 250]
Division: [0.1 0.1 0.1 0.1 0.1]

Scaler operation 

Exponetial: [ 100  400  900 1600 2500]
Floor division: [ 5 10 15 20 25]
Modulus: [1 2 0 1 2]


In [8]:
# The dot product or (scalar product) of two vectors a and b is a scalar value calculated by multipling corresponding
#components and summing them: 
# np.dot
# working with dot product

matrix_1 = np.array([
    [1, 2],
    [3, 4]

])

matrix_2 = np.array([
    [5, 6],
    [7, 8]
])

dot_product = np.dot(matrix_1, matrix_2)
multiply_arr = np.array(matrix_1 * matrix_2)

print("Dot Production output: \n", dot_product)
print("Multiply arr output: \n", multiply_arr)

Dot Production output: 
 [[19 22]
 [43 50]]
Multiply arr output: 
 [[ 5 12]
 [21 32]]


In [9]:
#4/5/2026 Monday
#Analyzing 30 days fittness tracker dataset 
#Pandas is used because
# np.arange is used to generate number within the range given 
# np.random.randint is used to generate random number within the range given 
# np.random.seed(42) is used for reproducibility (i.e to get thesame value with other people)
np.random.seed(42) #reproducibility
days = np.arange(1, 31)
steps = np.random.randint(3456, 12034, size=30)

print("Days:", days)
print("\nSteps:", steps)
                        

Days: [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30]

Steps: [10726  4316  8846  8647  9190  9721  3922  7882  9034 11778  5141  4225
 10405  5889  8767  8507  9876  4640  8011  6841  9852  6014 11305  5503
  6203  3645  6190  6461  8114  5355]


In [17]:
# Pandas is use to convert the output to rows and columns using "pd.DataFrame"
#Data frame is used to structure dataset into rows and columns
 
data = pd.DataFrame({
    "Days" :days, 
    "Steps" : steps
})

data

,Days,Steps
0,1,10726
1,2,4316
2,3,8846
3,4,8647
4,5,9190
5,6,9721
6,7,3922
7,8,7882
8,9,9034
9,10,11778


In [19]:
#basic stats 
print("Average step:" , np.mean(steps))
print("Median step:" , np.median(steps))
print("Minimum step:" , np.min(steps))
print("Maximum step:" , np.max(steps))
print("Step variability:" , np.std(steps))


Average step: 7500.2
Median step: 7946.5
Minimum step: 3645
Maximum step: 11778
Step variability: 2309.770715893679


In [20]:
#goal analysis step exceed 10,000
# where is used to filter out some part of the dataset
# [0] + 1 is added, so that the index number can be thesame as serial number

goal = np.where(steps > 10000)[0] + 1
print(goal)

[ 1 10 13 23]


In [13]:
goal_btw_1 = np.where((steps >= 3645) & (steps <= 7500)) [0] + 1
print(goal_btw_1)

[ 2  7 11 12 14 18 20 22 24 25 26 27 28 30]


In [14]:
goal_btw_2 = np.where((steps >= 8000) & (steps <= 11778)) [0] + 1
print(goal_btw_2)

[ 1  3  4  5  6  9 10 13 15 16 17 19 21 23 29]


In [15]:
# 5/5/2026 Tuesday
#weekly analysis (split into 4 weeks)

weeks = np.array_split(steps, 4)

print(weeks)

print("Week total steps:")
week_total = [print(week.sum()) for week in weeks]

print("\nWeek average steps")
week_avg = [print(week.mean()) for week in weeks]

[array([10726,  4316,  8846,  8647,  9190,  9721,  3922,  7882],
      dtype=int32), array([ 9034, 11778,  5141,  4225, 10405,  5889,  8767,  8507],
      dtype=int32), array([ 9876,  4640,  8011,  6841,  9852,  6014, 11305], dtype=int32), array([5503, 6203, 3645, 6190, 6461, 8114, 5355], dtype=int32)]
Week total steps:
63250
63746
56539
41471

Week average steps
7906.25
7968.25
8077.0
5924.428571428572


In [21]:
#detect anomaly/ outliers
mean_step = np.mean(steps)
std_step = np.std(steps)

lower_bond = mean_step - 2 * std_step
upper_bond = mean_step + 2 * std_step

print(lower_bond)
print(upper_bond)

#To round-up decimal point => round()
print(round(lower_bond, 2))
print(round(upper_bond))

outliers = np.where((steps < lower_bond) | (steps > upper_bond))[0] + 1

print("Anomaly / Outliers detected:", outliers)

2880.6585682126415
12119.741431787359
2880.66
12120
Anomaly / Outliers detected: []


The average daily steps is 7,50, meaning that on a typical day the person walk around seventy-five hundred steps.
    
The median is 7,946, a bit higher than the mean, which suggests most days are on the more active side.
    
The lowest day is 3,645 steps, likely a low-movement day, while the highest day reaches 11,778 marking the most active day of the month. 
    
The standard deviation is 2,309, showing that the activity varies a lot, some days are very active, others much lighter.